### ISSUE TITLE
Update Syslog source to accept non UTF-8 encoding in syslog message

### ISSUE THREAD

**Issue by Neko-Follower** [vectordotdev/vector#20462]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Vector drops logs when encounter a syslog message with non UTF-8 characters. Can you add an option to replace non utf-8 characters with U+FFFD or allow passing non-UTF8 text as-is like a Promtail do.

### Version

0.37.1-distroless-static

### Example Data

2024-05-08T07:35:43.209847Z DEBUG source{component_kind="source" component_id=rsyslog component_type=syslog}:connection{peer_addr=172.22.0.4:44600}: vector::sources::util::net::tcp: Accepted a new connection. peer_addr=172.22.0.4:44600

2024-05-08T07:35:44.293974Z ERROR source{component_kind="source" component_id=rsyslog component_type=syslog}:connection{peer_addr=172.22.0.4:44594}: vector::internal_events::codecs: Failed framing bytes. error=Unable to decode input as UTF8 error_code="decoder_frame" error_type="parser_failed" stage="processing" internal_log_rate_limit=true

2024-05-08T07:35:44.294029Z ERROR source{component_kind="source" component_id=rsyslog component_type=syslog}:connection{peer_addr=172.22.0.4:44594}: vector::internal_events::codecs: Internal log [Failed framing bytes.] is being suppressed to avoid flooding.

**Comment by jszwedko** [comment]
Agreed, this could be modeled like the existing `decoding.codec.json.lossy` option which replaces invalid UTF-8 characters.

**Comment by jszwedko** [comment]
We'd be happy to see a PR for this if someone is motivated! It should be a relatively straightforward change.

**Comment by osas1111** [comment]
I have the same problem with Fluent source

**Comment by kevinmingtarja** [comment]
Hi, I'm interested in contributing to this!

**Comment by jszwedko** [comment]
> Hi, I'm interested in contributing to this!

Great! We'd be happy to review a PR. You can see https://github.com/vectordotdev/vector/pull/17628 as an example of when it was added to the JSON decoder.

**Comment by kevinmingtarja** [comment]
Hi @jszwedko, just to confirm, it seems like the `lossy` option for syslog has been added in this PR #17680.

Is there anything else missing from that PR that could be causing this bug?

I wrote a unit test for `/lib/codecs/src/decoding/format/syslog.rs` and was able to verify that `SyslogDeserializer::default().parse()` does replace the non UTF-8 characters with the replacement character.

**Comment by jszwedko** [comment]
Ah, and so it was. I forgot this issue is about the `syslog` source rather than the `syslog` decoder. I think we still need to add the option to [the source](https://vector.dev/docs/reference/configuration/sources/syslog/).

**Comment by drawks** [comment]
Defining your source in the config using the `socket` type instead of the `syslog` type allows you to get the desired behavior.

Something like this
```diff
 [sources.syslog]
-type = "syslog"
+type = "socket"
 address = "127.0.0.1:31337"
 mode = "tcp"
 max_length = 20_000_000
+decoding.codec = "syslog"
+decoding.syslog.lossy = true  # escape non-utf8 bytes
```

**Comment by mezgerj** [comment]
I took a shot at this one in https://github.com/vectordotdev/vector/pull/24657
